src/test/java/io/cdap/plugin/salesforce/etl/BaseSalesforceBatchSinkETLTest.java

```diff
@@ -0,0 +1,219 @@
+import io.cdap.cdap.proto.ProgramRunStatus;
+import io.cdap.cdap.proto.artifact.AppRequest;
+import io.cdap.cdap.proto.id.ApplicationId;
+import io.cdap.cdap.proto.id.ArtifactId;
+import io.cdap.cdap.proto.id.NamespaceId;
+import io.cdap.cdap.test.ApplicationManager;
+import io.cdap.cdap.test.DataSetManager;
+import io.cdap.cdap.test.TestConfiguration;
+import io.cdap.cdap.test.WorkflowManager;
+import io.cdap.plugin.salesforce.plugin.sink.batch.SalesforceBatchSink;
+import io.cdap.plugin.salesforce.plugin.sink.batch.SalesforceSinkConfig;
+import io.cdap.plugin.salesforce.plugin.sink.batch.StructuredRecordToCSVRecordTransformer;
+import org.junit.Before;
+import org.junit.BeforeClass;
+import org.junit.ClassRule;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+/**
+ * {@inheritDoc}
+ */
+public abstract class BaseSalesforceBatchSinkETLTest extends BaseSalesforceETLTest {
+  @ClassRule
+  public static final TestConfiguration CONFIG = new TestConfiguration("explore.enabled", false);
+  private static final ArtifactSummary APP_ARTIFACT = new ArtifactSummary("data-pipeline", "3.2.0");
+
+  private static final String REFERENCE_NAME = "SalesforceBatchSink-output";
+  private int deployCounter;
+
+  @BeforeClass
+  public static void setupTestClass() throws Exception {
+    ArtifactId parentArtifact = NamespaceId.DEFAULT.artifact(APP_ARTIFACT.getName(), APP_ARTIFACT.getVersion());
+
+    // add the artifact and mock plugins
+    setupBatchArtifacts(parentArtifact, DataPipelineApp.class);
+
+    // add our plugins artifact with the artifact as its parent.
+    // this will make our plugins available.
+    addPluginArtifact(NamespaceId.DEFAULT.artifact("example-plugins", "1.0.0"),
+                      parentArtifact,
+                      SalesforceBatchSink.class
+    );
+  }
+
+  @Before
+  public void init() {
+    deployCounter = 0;
+  }
+
+  /**
+   * Assert that records provided all exist in Salesforce with the exact fields.
+   * If not throw and exception.
+   *
+   * @param sObject sObject name
+   * @param inputRecords list of records which are checked exist in Salesforce
+   * @throws ConnectionException occurs due to failure to connect to Salesforce API
+   */
+  protected void assertRecordsCreated(String sObject, List<StructuredRecord> inputRecords) throws ConnectionException {
+    List<String> errors = new ArrayList<>(inputRecords.size());
+    for (StructuredRecord record : inputRecords) {
+      String query = getQueryByRecord(sObject, record);
+      SObject[] sObjects = partnerConnection.query(query).getRecords();
+      if (sObjects.length == 0) {
+        String errorMessage = String.format("Not found record='%s'. Query='%s'", record, query);
+        errors.add(errorMessage);
+      }
+      addToCleanUpList(sObjects);
```

Comment:
can we remove this side effect from assertion method?

Reply:
I thought about this as well. But I could not figure out how we can gather all the ids which were successfully retrieved and remove them, before failing the method on objects which were not retrieved.

Reply:
We could create a custom excecption and add information about sObjects which were parsed there. And than catch it in subclass. But than again it's also hacky to add that non-related info to an exception. And also it makes the subclass calls more dirty.

Reply:
done